merge/update.go

```diff
@@ -250,33 +250,54 @@ func (s *Updater) addBackOwnedItems(merged, pruned *typed.TypedValue, managedFie
 		}
 		managedAtVersion[managerSet.APIVersion()] = managedAtVersion[managerSet.APIVersion()].Union(managerSet.Set())
 	}
-	for version, managed := range managedAtVersion {
-		merged, err = s.Converter.Convert(merged, version)
+	// Add back owned items at pruned version first to avoid conversion failure
+	// caused by pruned fields which are required for conversion.
+	prunedVersion := fieldpath.APIVersion(*pruned.TypeRef().NamedType)
```

Comment:
I don't think that's correct, that's only an assumption we make in tests (where we use the name of the type as it's version), but once in kubernetes, the names of the type don't have a one-to-one mapping with their version.

Reply:
You could pass-in the version as a parameter I guess, since you have it in the `prune` function: `lastSet.APIVersion()`

Reply:
Can we add a test that will catch this?

Reply:
We can, but the "internal/fixture/state" strongly expects the type to be equivalent to the version (it does so as a convenience so that the test framework doesn't have to know how to extract the version from the object, /shrug).

We could write a test, at least one, that doesn't depend on the "internal/fixture/state" and uses incompatible version/typename to detect these issues.

Reply:
> write a test, at least one, that doesn't depend on the "internal/fixture/state" and uses incompatible version/typename to detect these issues

SGTM!

Reply:
I'm giving it a quick try right now